File: sample-projects/03-simple-document/03-simple-document/PMDataScanner.m
//
//  PMDataScanner.m
//  canari
//
//  Created by Pierre Molinaro on 09/11/07.
//  Copyright 2007 __MyCompanyName__. All rights reserved.
//
//---------------------------------------------------------------------------*

#import "PMDataScanner.h"
#import "easy-bindings-utilities.h"
#import "PMAllocationDebug.h"

//#define EASY_BINDINGS_DEBUG

//---------------------------------------------------------------------------*

@implementation PMDataScanner

//---------------------------------------------------------------------------*

- (void) openProgressWindowWithTitle: (NSString *) inTitle {
  const NSRect visibleFrame = [[NSScreen mainScreen] visibleFrame] ;
  const double windowWidth = 400.0 ;
  const double windowHeight = 65.0 ;
  const NSRect windowRect = {
    {NSMidX (visibleFrame) - windowWidth / 2.0, NSMidY (visibleFrame) - windowHeight / 2.0},
    {windowWidth, windowHeight}
  } ;
  mProgressWindow = [[NSWindow alloc]
    initWithContentRect:windowRect
    styleMask:NSTitledWindowMask
    backing:NSBackingStoreBuffered
    defer:NO
  ] ;
  [mProgressWindow setExcludedFromWindowsMenu:YES] ;
  [mProgressWindow setTitle:@"Progress"] ;
  const NSRect contientViewRect = [[mProgressWindow contentView] frame] ;
//--- Add comment text
  const NSRect ts_r = {{25.0, 30.0}, {NSMaxX (contientViewRect) - 40.0, 20.0}} ;
  NSTextField * ts = [[NSTextField alloc] initWithFrame:ts_r] ;
  [ts setFont:[NSFont boldSystemFontOfSize:[NSFont smallSystemFontSize]]] ;
  [ts setStringValue:[NSString stringWithFormat:@"Opening %@…", inTitle]] ;
  [ts setBezeled:NO] ;
  [ts setBordered:NO] ;
  [ts setEditable:NO] ;
  [ts setDrawsBackground:NO] ;
  [[mProgressWindow contentView] addSubview:ts] ;
  macroReleaseSetToNil (ts) ;
//--- Add progress indicator
  const NSRect ps_r = {{20.0, 10.0}, {NSMaxX (contientViewRect) - 40.0, 20.0}} ;
  mProgressIndicator = [[NSProgressIndicator alloc] initWithFrame:ps_r] ;
  [mProgressIndicator setIndeterminate:YES] ;
  [[mProgressWindow contentView] addSubview:mProgressIndicator] ;
//---
  [mProgressIndicator setMinValue:0.0] ;
  [mProgressIndicator setMaxValue: (double) [mData length]] ;
  [mProgressIndicator setDoubleValue: (double) mReadIndex] ;
  [mProgressIndicator setIndeterminate:NO] ;
  [mProgressIndicator display] ;
  [mProgressWindow makeKeyAndOrderFront:nil] ;
}

//---------------------------------------------------------------------------*

- (void) orderOutProgressWindow {
  [mProgressWindow close] ;
  mProgressWindow = nil ;
}

//---------------------------------------------------------------------------*

- (void) configureWithData: (NSData *) inData {
    mReadIndex = 0 ;
    mData = inData ;
    mReadOk = YES ;
    mExpectedBytes = [NSMutableData new] ;
}

//---------------------------------------------------------------------------*

+ (PMDataScanner *) scannerWithData: (NSData *) inData
                    displayProgressWindowTitle: (NSString *) inTitle {
  PMDataScanner * scanner  = [PMDataScanner new] ;
  [scanner configureWithData:inData] ;
  if (nil != inTitle) {
    [scanner openProgressWindowWithTitle:inTitle] ;
  }
  macroAutorelease (scanner) ;
  return scanner ;
}

//---------------------------------------------------------------------------*
//                                                                           *
//                           dealloc                                         *
//                                                                           *
//---------------------------------------------------------------------------*

#ifdef PM_COCOA_DEBUG
  - (void) dealloc {
    macroNoteObjectDeallocation ;
    macroSuperDealloc ;
  }
#endif

//---------------------------------------------------------------------------*
//                                                                           *
//                           dealloc                                         *
//                                                                           *
//---------------------------------------------------------------------------*

- (id) init {
  self = [super init] ;
  if (self) {
    macroNoteObjectAllocation ;
  }
  return self ;
}

//---------------------------------------------------------------------------*

- (void) updateProgressIndicator {
  [mProgressIndicator setDoubleValue: (double) mReadIndex] ;
  [mProgressIndicator display] ;
}

//---------------------------------------------------------------------------*

- (void) ignoreBytes: (NSUInteger) inLengthToIgnore {
  if (mReadOk) {
    mReadIndex += inLengthToIgnore ;
    [self updateProgressIndicator] ;
  }
}

//---------------------------------------------------------------------------*

- (BOOL) testAcceptByte: (UInt8) inByte {
  BOOL result = mReadOk ;
  if (result) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      result = bytes [mReadIndex] == inByte ;
      if (result) {
        mReadIndex ++ ;
        [mExpectedBytes setLength:0] ;
      }else{
        [mExpectedBytes appendBytes:& inByte length:1] ;
      }
    }
  }
  [self updateProgressIndicator] ;
  return result ;
}

//---------------------------------------------------------------------------*

- (BOOL) testAcceptFromByte: (UInt8) inLowerBound
         toByte: (UInt8) inUpperBound
         value: (UInt8 *) outValuePtr {
  BOOL result = mReadOk ;
  if (result) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      result = (bytes [mReadIndex] >= inLowerBound) && (bytes [mReadIndex] <= inUpperBound) ;
      if (result) {
        if (outValuePtr != NULL) {
          * outValuePtr = bytes [mReadIndex] ;
        }
        mReadIndex ++ ;
        [mExpectedBytes setLength:0] ;
      }else{
        for (NSUInteger i=inLowerBound ; i<=inUpperBound ; i++) {
          [mExpectedBytes appendBytes:& i length:1] ;
        }
      }
    }
  }
  [self updateProgressIndicator] ;
  return result ;
}

//---------------------------------------------------------------------------*

- (void) acceptRequiredByte: (UInt8) inByte
         sourceFile: (const char *) inSourceFile {
  if (mReadOk) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      if (bytes [mReadIndex] == inByte) {
        mReadIndex ++ ;
        [mExpectedBytes setLength:0] ;
      }else{
        NSMutableString * message = [NSMutableString new] ;
        const UInt8 * expectedBytes = [mExpectedBytes bytes] ;
        for (NSUInteger i=0 ; i<[mExpectedBytes length] ; i++) {
          [message appendFormat:@"0x%02hhx, ", expectedBytes  [i]] ;
        }
        NSLog (@"%s: invalid current byte (0x%02x): expected bytes:%@0x%02x", inSourceFile, bytes [mReadIndex], message, inByte) ;
        macroReleaseSetToNil (message) ;
        mReadOk = NO ;
      }
    }
  }
  [self updateProgressIndicator] ;
}

//---------------------------------------------------------------------------*

- (UInt8) parseByte {
  UInt8 result = 0 ;
  if (mReadOk) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
     }else{
      const UInt8 * bytes = [mData bytes] ;
      result = bytes [mReadIndex] ;
      mReadIndex ++ ;
    }
  }
  [self updateProgressIndicator] ;
  return result ;
}

//---------------------------------------------------------------------------*

- (UInt64) parseAutosizedUnsignedInteger {
  UInt64 result = 0 ;
  const UInt8 * bytes = [mData bytes] ;
  NSUInteger shift = 0 ;
  BOOL loop = YES ;
  while (loop && mReadOk) {
    if (mReadIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
    }else{
      const UInt64 w = bytes [mReadIndex] & 0x7F ;
      result |= (UInt64) (w << shift) ;
      shift += 7 ;
      loop = (bytes [mReadIndex] & 0x80) !=0 ;
      mReadIndex ++ ;
    }
  }
  [self updateProgressIndicator] ;
  return result ;
}

//---------------------------------------------------------------------------*

- (NSString *) parseAutosizedString {
  NSString * result = nil ;
  const UInt8 * bytes = [mData bytes] ;
  NSUInteger beyondStringIndex = mReadIndex ;
  BOOL loop = YES ;
  while (loop && mReadOk) {
    if (beyondStringIndex >= [mData length]) {
       NSLog (@"Read beyond end of data") ;
       mReadOk = NO ;
    }else{
      loop = bytes [beyondStringIndex] != 0 ;
      beyondStringIndex ++ ;
    }
  }
  if (mReadOk) {
    result = [NSString stringWithUTF8String:(const char *) & bytes [mReadIndex]] ;
    mReadIndex = beyondStringIndex ;
  }
  [self updateProgressIndicator] ;
  return result ;
}

//---------------------------------------------------------------------------*

- (NSData *) parseAutosizedData {
  NSData * result = nil ;
  if (mReadOk) {
    const NSUInteger dataLength = [self parseAutosizedUnsignedInteger] ;
    if ((mReadIndex + dataLength) >= [mData length]) {
      NSLog (@"Read beyond end of data") ;
      mReadOk = NO ;
    }else{
      result = [mData subdataWithRange:NSMakeRange (mReadIndex, dataLength)] ;
      mReadIndex += dataLength ;
    }
  }
  [self updateProgressIndicator] ;
  return result ;
}


//---------------------------------------------------------------------------*

- (void) ignoreAutosizedData {
  if (mReadOk) {
    const NSUInteger dataLength = [self parseAutosizedUnsignedInteger] ;
    [self ignoreBytes:dataLength] ;
  }
}

//---------------------------------------------------------------------------*

- (BOOL) ok {
  return mReadOk ;
}

//---------------------------------------------------------------------------*

@end

File: sample-projects-swift/03-simple-document-swift/03-simple-document/PMDataScanner.swift
//
//  PMDataScanner.m
//  canari
//
//  Created by Pierre Molinaro on 09/11/07.
//  Copyright 2007 __MyCompanyName__. All rights reserved.
//
//---------------------------------------------------------------------------*

import Cocoa

//---------------------------------------------------------------------------*

@objc(PMDataScanner) class PMDataScanner : NSObject {
  var mData : NSData
  var mReadIndex : Int = 0
  var mReadOk : Bool = true
  var mExpectedBytes : Array<UInt8> = []
  var mProgressWindow : NSWindow?
  var mProgressIndicator : NSProgressIndicator?

  //---------------------------------------------------------------------------*
  //  init                                                                     *
  //---------------------------------------------------------------------------*

  init (data: NSData,
        displayProgressWindowTitle: String?) {
    mData = data
    super.init ()
    noteObjectAllocation (self)
    if (nil != displayProgressWindowTitle) {
      openProgressWindowWithTitle (displayProgressWindowTitle!)
    }
  }

  //---------------------------------------------------------------------------*
  //  deinit                                                                   *
  //---------------------------------------------------------------------------*

  deinit {
    mProgressWindow?.orderOut (self)
    noteObjectDeallocation (self)
  }

  //---------------------------------------------------------------------------*
  //  openProgressWindowWithTitle                                              *
  //---------------------------------------------------------------------------*

  func openProgressWindowWithTitle (inTitle : String) {
    let visibleFrame = NSScreen.mainScreen ().visibleFrame ()
    let windowWidth = 400.0
    let windowHeight = 65.0
    let windowRect = NSRect (
      x:NSMidX (visibleFrame) - windowWidth / 2.0,
      y:NSMidY (visibleFrame) - windowHeight / 2.0,
      width:windowWidth,
      height:windowHeight
    )
    mProgressWindow = NSWindow (
      contentRect:windowRect,
      styleMask:NSTitledWindowMask,
      backing:NSBackingStoreBuffered,
      defer:false
    )
    mProgressWindow!.setExcludedFromWindowsMenu (true)
    mProgressWindow!.setTitle ("Progress")
    let contientViewRect : NSRect = mProgressWindow!.contentView().frame ()
  //--- Add comment text
    let ts_r = NSRect (
      x:25.0,
      y:30.0,
      width:NSMaxX (contientViewRect) - 40.0,
      height:20.0
    )
    var ts = NSTextField (frame:ts_r)
    ts.setFont (NSFont.boldSystemFontOfSize (NSFont.smallSystemFontSize ()))
    ts.setStringValue (NSString (format:"Opening %@…", inTitle))
    ts.setBezeled (false)
    ts.setBordered (false)
    ts.setEditable (false)
    ts.setDrawsBackground (false)
    mProgressWindow!.contentView ().addSubview (ts)
  //--- Add progress indicator
    let ps_r = NSRect (
      x:20.0,
      y:10.0,
      width:NSMaxX (contientViewRect) - 40.0,
      height: 20.0
    )
    mProgressIndicator = NSProgressIndicator (frame:ps_r)
    mProgressIndicator!.setIndeterminate (true)
    mProgressWindow!.contentView ().addSubview (mProgressIndicator)
  //---
    mProgressIndicator!.setMinValue (0.0)
    mProgressIndicator!.setMaxValue (Double (mData.length ()))
    mProgressIndicator!.setDoubleValue (Double (mReadIndex))
    mProgressIndicator!.setIndeterminate (false)
    mProgressIndicator!.display ()
  //---
    mProgressWindow!.makeKeyAndOrderFront (nil)
  }

  //---------------------------------------------------------------------------*
  //  updateProgressIndicator                                                  *
  //---------------------------------------------------------------------------*

  func updateProgressIndicator () {
    mProgressIndicator?.setDoubleValue (Double (mReadIndex))
    mProgressIndicator?.display ()
  }

  //---------------------------------------------------------------------------*
  //  ignoreBytes                                                              *
  //---------------------------------------------------------------------------*

  func ignoreBytes (inLengthToIgnore : Int) {
    if mReadOk {
      mReadIndex += inLengthToIgnore ;
      updateProgressIndicator ()
    }
  }

  //---------------------------------------------------------------------------*
  //  testAcceptByte                                                           *
  //---------------------------------------------------------------------------*
  // http://stackoverflow.com/questions/24067085/pointers-pointer-arithmetic-and-raw-data-in-swift

  func testAcceptByte (inByte : UInt8) -> Bool {
    var result = mReadOk
    if result {
      if mReadIndex >= mData.length () {
         NSLog ("Read beyond end of data")
         mReadOk = false
       }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData()).memory
     //   let ptr = offsetPointer (mData.bytes, CInt(mReadIndex))
     //   var array : COpaquePointer = mData.bytes ()
     //   var byteArray : CConstPointer<UInt8> = CConstPointer<UInt8> (mData.bytes ())
        result = byte == inByte
        if result {
          mReadIndex += 1
          mExpectedBytes = []
        }else{
          mExpectedBytes += inByte
        }
      }
    }
    updateProgressIndicator ()
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  testAcceptFromByte                                                       *
  //---------------------------------------------------------------------------*

  func testAcceptFromByte (lowerBound: UInt8,
                           upperBound: UInt8,
                           inout value:UInt8) -> Bool {
    var result = mReadOk
    if result {
      if mReadIndex >= mData.length () {
         NSLog ("Read beyond end of data")
         mReadOk = false
       }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData()).memory
        result = (byte >= lowerBound) && (byte <= upperBound) ;
        if (result) {
          value = byte
          mReadIndex += 1
          mExpectedBytes = []
        }else{
          for i in lowerBound ..< upperBound + 1 {
            mExpectedBytes += i
          }
        }
      }
    }
    updateProgressIndicator ()
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  acceptRequiredByte                                                       *
  //---------------------------------------------------------------------------*

  func acceptRequiredByte (inByte : UInt8,
                           sourceFile: String) {
    if mReadOk {
      if mReadIndex >= mData.length () {
         NSLog ("Read beyond end of data")
         mReadOk = false
      }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData()).memory
        if (byte == inByte) {
          mReadIndex += 1
          mExpectedBytes = []
        }else{
          var message = ""
          for b in mExpectedBytes {
            message += NSString (format:"0x%02hhx, ", b)
          }
          NSLog ("%s: invalid current byte (0x%02x): expected bytes:%@0x%02x", sourceFile, byte, message, inByte) ;
          mReadOk = false
        }
      }
    }
    updateProgressIndicator ()
  }

  //---------------------------------------------------------------------------*
  //  parseByte                                                                *
  //---------------------------------------------------------------------------*

  func parseByte () -> UInt8 {
    var result : UInt8 = 0
    if mReadOk {
      if mReadIndex >= mData.length() {
         NSLog ("Read beyond end of data")
         mReadOk = false
       }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        result = UnsafePointer<UInt8> (byteAsData()).memory
        mReadIndex += 1
      }
    }
    updateProgressIndicator ()
    return result
  }

  //---------------------------------------------------------------------------*
  //  parseAutosizedUnsignedInteger                                            *
  //---------------------------------------------------------------------------*

  func parseAutosizedUnsignedInteger () -> UInt {
    var result : UInt = 0
    var shift : UInt = 0
    var loop = true
    while loop && mReadOk {
      if mReadIndex >= mData.length () {
         NSLog ("Read beyond end of data")
         mReadOk = false
      }else{
        let byteAsData = mData.subdataWithRange (NSMakeRange(mReadIndex, sizeof(UInt8))).bytes
        var byte = UnsafePointer<UInt8> (byteAsData()).memory
        let w : UInt = UInt (byte) & 0x7F
        result |= (w << shift)
        shift += 7
        loop = (byte & 0x80) != 0
        mReadIndex += 1
      }
    }
    updateProgressIndicator ()
    return result ;
  }

  //---------------------------------------------------------------------------*
  //  parseAutosizedData                                                       *
  //---------------------------------------------------------------------------*

  func parseAutosizedData () -> NSData {
  var result = NSData ()
  if mReadOk {
    let dataLength : Int = Int (parseAutosizedUnsignedInteger ())
    if (mReadIndex + dataLength) >= mData.length () {
      NSLog ("Read beyond end of data")
      mReadOk = false
    }else{
      result = mData.subdataWithRange (NSMakeRange (mReadIndex, dataLength))
      mReadIndex += dataLength
    }
  }
  updateProgressIndicator ()
  return result ;
}

  //---------------------------------------------------------------------------*
  //  ok                                                                       *
  //---------------------------------------------------------------------------*

  func ok () -> Bool {
    return mReadOk
  }

}

//---------------------------------------------------------------------------*
